compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -259,44 +260,105 @@ impl<'hir> LoweringContext<'_, 'hir> {
                         self_param_id: pat_node_id,
                     };
                     self_resolver.visit_block(block);
-                    let block = this.lower_block(block, false);
-                    this.mk_expr(hir::ExprKind::Block(block, None), block.span)
+                    let mut block = this.lower_block_noalloc(block, false);
+                    if block.expr.is_none() {
+                        let pat_hir_id = this.lower_node_id(pat_node_id);
+                        let arg = &*this.arena.alloc(this.generate_arg(pat_hir_id, span));
+                        // Use `self` as a first argument if no expression is provided.
```

Comment:
"Omitted target expression" in my comment meant `reuse foo;`, not `reuse foo {}`.
In the second case the target expression is indeed a unit, the old desugaring was correct.